UsageStats: Add traces when sending usage stats

sakjur opened the pull request:
**What this PR does / why we need it**:

Waiting for #55198.

- Improves logging for usage stats.
- Introduces tracing usage stats running as a background job.
- Adds top level tracing for each collector (including including the name of the function which is called in the span name).

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #55456

**Special notes for your reviewer**:

sakjur committed:
UsageStats: Introduce tracing

marefr commented:
#55198 is merged :tada:

sakjur committed:
How to send a fake tracer?

sakjur committed:
Merge remote-tracking branch 'origin/main' into emil/202209/trace-usagestats

sakjur committed:
Use new contextual logger and style from tracing guidelines

sakjur committed:
Set status on failure and restructure the tracing

sakjur commented on pkg/infra/usagestats/service/usage_stats.go:
```diff
@@ -66,47 +73,75 @@ func (uss *UsageStats) gatherMetrics(ctx context.Context, metrics map[string]int
 //
 // Stubbable by tests.
-var sendUsageStats = func(uss *UsageStats, data *bytes.Buffer) {
-	go func() {
-		client := http.Client{Timeout: 5 * time.Second}
-		resp, err := client.Post(usageStatsURL, "application/json", data)
-		if err != nil {
-			uss.log.Error("Failed to send usage stats", "err", err)
-			return
-		}
-		if err := resp.Body.Close(); err != nil {
-			uss.log.Warn("Failed to close response body", "err", err)
-		}
-	}()
```
This is a notable change to get the entirety of the function inside of the span.

From what I could tell, the only reason to create a Go routine here is for a test, which could equally well be solved by creating that Go routine in the test.

sakjur replied:
> Or maybe, if for whatever reason a panic would happen it could crash the goroutine of the background job.

Good thinking, ran that scenario through my head and I don't think that behavior would change prior/after, as a panic would escalate all the way out and crash Grafana even if it happens in a Go routine? [Spec on handling panics](https://go.dev/ref/spec#Handling_panics) says:

> While executing a function F, an explicit call to panic or a [run-time panic](https://go.dev/ref/spec#Run_time_panics) terminates the execution of F. Any functions [deferred](https://go.dev/ref/spec#Defer_statements) by F are then executed as usual. Next, any deferred functions run by F's caller are run, and so on up to any deferred by the top-level function in the executing goroutine. At that point, the program is terminated and the error condition is reported, including the value of the argument to panic.

sakjur committed:
Merge branch 'main' into emil/202209/trace-usagestats

marefr commented on pkg/infra/usagestats/service/usage_stats.go:
```diff
@@ -66,47 +73,75 @@ func (uss *UsageStats) gatherMetrics(ctx context.Context, metrics map[string]int
+	defer span.End()
+
+	client := http.Client{Timeout: 5 * time.Second}
+	resp, err := client.Post(usageStatsURL, "application/json", data)
```
Maybe should include the `ctx` for the outgoing request? Should we inject trace headers for the outgoing request?

sakjur replied:
Are you able to link to an example of where that's been done?

sakjur replied:
```suggestion
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, usageStatsURL, data)
	req.Header.Set("Content-Type", "application/json")
	uss.tracer.Inject(ctx, req.Header, span)
	resp, err := client.Do(req)
```

Like this?

marefr replied:
https://github.com/grafana/grafana/blob/e714d750f2fc31ee50db4d347be8900bbef58b2a/pkg/api/pluginproxy/ds_proxy.go#L149

marefr replied:
https://github.com/grafana/grafana/blob/e714d750f2fc31ee50db4d347be8900bbef58b2a/pkg/tsdb/graphite/graphite.go#L155

marefr replied:
Yes, that looks correct :+1:

marefr replied:
Yes, could be risky to introduce since it's only used for datasources currently. Would suggest considering that in a separate PR together with tests

sakjur replied:
I kind of thought so, I like the idea of using it, but agree that we'll wait til a separate PR

sakjur committed:
Inject trace into HTTP request

Considered using httpclient, but decided against it this time to try to
keep this PR a bit streamlined.
